common/raw_hashtable.h

```diff
@@ -0,0 +1,1333 @@
+    // early.
+    CARBON_DCHECK(static_cast<uint64_t>(alloc_size) ==
+                  llvm::alignTo<alignof(EntryT)>(alloc_size));
+    return alloc_size;
+  }
+
+  auto metadata() const -> uint8_t* {
+    return reinterpret_cast<uint8_t*>(storage_);
+  }
+  auto entries() const -> EntryT* {
+    return reinterpret_cast<EntryT*>(reinterpret_cast<std::byte*>(storage_) +
+                                     EntriesOffset(alloc_size_));
+  }
+
+  ssize_t alloc_size_;
+  Storage* storage_;
+};
+
+// Implementation helper for defining a read-write base type for a hashtable
+// that type-erases any SSO buffer.
+//
+// A specific user-facing hashtable base type should derive using *`protected`*
+// inheritance from this type, and forward the implementation of its interface
+// to functions in this type.
+//
+// Other than the use of `protected` inheritance, the patterns for this type,
+// and how to build user-facing hashtable base types from it, mirror those of
+// `ViewImpl`. See its documentation for more details.
+template <typename InputKeyT, typename InputValueT, typename InputKeyContextT>
+class BaseImpl {
+ protected:
+  using KeyT = InputKeyT;
+  using ValueT = InputValueT;
+  using KeyContextT = InputKeyContextT;
+  using ViewImplT = ViewImpl<KeyT, ValueT, KeyContextT>;
+  using EntryT = typename ViewImplT::EntryT;
+
+  BaseImpl(int small_alloc_size, Storage* small_storage)
+      : small_alloc_size_(small_alloc_size) {
+    CARBON_CHECK(small_alloc_size >= 0);
+    Construct(small_storage);
+  }
+  // Only used for copying and moving, and leaves storage uninitialized.
+  BaseImpl(ssize_t alloc_size, int growth_budget, int small_alloc_size)
+      : view_impl_(alloc_size, nullptr),
+        growth_budget_(growth_budget),
+        small_alloc_size_(small_alloc_size) {}
+  ~BaseImpl();
+
+  // NOLINTNEXTLINE(google-explicit-constructor): Designed to implicitly decay.
+  operator ViewImplT() const { return view_impl(); }
+
+  auto view_impl() const -> ViewImplT { return view_impl_; }
+
+  // Looks up the provided key in the hashtable. If found, returns a pointer to
+  // that entry and `false`.
+  //
+  // If not found, will locate an empty entry for inserting into, set the
+  // metadata for that entry, and return a pointer to the entry and `true`. When
+  // necessary, this will grow the hashtable to cause there to be sufficient
+  // empty entries.
+  template <typename LookupKeyT>
+  auto InsertImpl(LookupKeyT lookup_key,
+                  KeyContextT key_context = KeyContextT())
+      -> std::pair<EntryT*, bool>;
+
+  // Looks up the entry in the hashtable, and if found destroys the entry and
+  // returns `true`. If not found, returns `false`.
+  //
+  // Does not release any memory, just leaves a tombstone behind so this entry
+  // cannot be found and the slot can in theory be re-used.
+  template <typename LookupKeyT>
+  auto EraseImpl(LookupKeyT lookup_key, KeyContextT key_context = KeyContextT())
```

Comment:
Any reason the raw versions of these operations should have default values? Seems like that would just invited accidents where the key context isn't passed through.